File: LeetCode/1379. Find a Corresponding Node of a Binary Tree in a Clone of That Tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

/**
 * Runtime: 508 ms
 * Memory Usage: 16.1 MB 
 */
class Solution {
public:
    TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
        deque<TreeNode*> que;
        que.push_back(cloned);
        while (!que.empty()) {
            int size = que.size();
            for (int i = 0; i < size; ++ i) {
                auto node = que.front(); que.pop_front();
                if (node->val == target->val) {
                    return node;
                }
                if (node->left != nullptr) que.push_back(node->left);
                if (node->right != nullptr) que.push_back(node->right);
            }
        }
        return nullptr;
    }
};

File: LeetCode/1379. Find a Corresponding Node of a Binary Tree in a Clone of That Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
Runtime: 1248 ms
Memory Usage: 25.3 MB
"""
class Solution:
    def getTargetCopy(self, original: TreeNode, cloned: TreeNode, target: TreeNode) -> TreeNode:
        self.ans = None
        def dfs(node):
            if not node:
                return
            if node.val == target.val:
                self.ans = node
                return
            dfs(node.left)
            dfs(node.right)
        dfs(cloned)
        return self.ans
